## /pr-review — CHANGES REQUIRED

**Result:** 4 of 6 phases PASS. Phase 3 (Testing) and Phase 5 (CI) BLOCKED by a single regression introduced by this PR.

### Phase Results
- Phase 1 (PR Overview): PASS
- Phase 2 (Security & Code Quality): PASS — no central provider violations, no secrets, nil guards correct, backward compat preserved
- Phase 3 (Testing): **FAIL — blocking**
- Phase 4 (Documentation): PASS — README, ADR-003, and godocs are thorough
- Phase 5 (CI Verification): **FAIL — blocking** (`unit-tests` and `Build Gate` failing, downstream of the same test regression)
- Phase 6 (Approval): BLOCKED

---

## Blocking Issue (single root cause)

**File:** `pkg/testing/storage/fixtures.go`

This PR changed `pkg/testing/storage_helper.go` to register the `flatfile` and `sqlite` providers (via blank imports) instead of `git`. That registration is **global** — any package that imports `pkg/testing` now has `flatfile` and `sqlite` in `interfaces.GetRegisteredProviderNames()`.

`pkg/testing/storage/fixtures.go` was **not** updated to match. Its `NewStorageTestFixture` only builds configs for `"git"` and `"database"`. When `TestStorageTestFixture_AllProviders` iterates registered providers, it finds `"flatfile"` and `"sqlite"` with no matching config entry and fails at `fixtures.go:202`:

```
--- FAIL: TestStorageTestFixture_AllProviders/provider_flatfile (0.00s)
    fixtures.go:292: Test configuration not found for provider 'flatfile'
--- FAIL: TestStorageTestFixture_AllProviders/provider_sqlite (0.00s)
    fixtures.go:292: Test configuration not found for provider 'sqlite'
FAIL  github.com/cfgis/cfgms/pkg/testing/storage  0.101s
```

This is the only blocking finding. Everything else is sound.

---

## Required Fix

### 1. Edit `pkg/testing/storage/fixtures.go`

**(a) Add two new setup methods**, modeled on the existing `setupGitConfig` (lines 109-126). Place them alongside the other `setup*Config` methods.

The config keys must match what each provider expects (verified in the provider source):
- **flatfile** requires `"root"` (string) — root directory path. See `pkg/storage/providers/flatfile/plugin.go:100`.
- **sqlite** requires `"path"` (string) — database file path. See `pkg/storage/providers/sqlite/plugin.go:134`. **Do NOT use `:memory:`** — the project README explicitly warns against it in parallel tests. Use a real temp file.

```go
// setupFlatfileConfig creates a flatfile provider configuration backed by a temp dir.
func (f *StorageTestFixture) setupFlatfileConfig(t *testing.T) {
	flatfileDir := filepath.Join(f.TempDir, "flatfile-storage")
	err := os.MkdirAll(flatfileDir, 0755)
	require.NoError(t, err, "Failed to create flatfile storage directory")

	f.Configs["flatfile"] = &StorageTestConfig{
		Provider: "flatfile",
		Config: map[string]interface{}{
			"root": flatfileDir,
		},
		TempDir: flatfileDir,
	}
}

// setupSQLiteConfig creates a sqlite provider configuration backed by a temp file.
// Uses a real file path (not :memory:) to keep each sub-store on the same DB and
// to avoid parallel-test issues flagged in pkg/storage/interfaces/README.md.
func (f *StorageTestFixture) setupSQLiteConfig(t *testing.T) {
	sqliteDir := filepath.Join(f.TempDir, "sqlite-storage")
	err := os.MkdirAll(sqliteDir, 0755)
	require.NoError(t, err, "Failed to create sqlite storage directory")

	f.Configs[\"sqlite\"] = &StorageTestConfig{
		Provider: \"sqlite\",
		Config: map[string]interface{}{
			\"path\": filepath.Join(sqliteDir, \"cfgms-test.db\"),
		},
		TempDir: sqliteDir,
	}
}
```

**(b) Call both from `NewStorageTestFixture`** (after the existing `setupDatabaseConfig` call at line 104):

```go
fixture.setupGitConfig(t)
fixture.setupDatabaseConfig(t)
fixture.setupFlatfileConfig(t)   // ADD
fixture.setupSQLiteConfig(t)     // ADD
```

### 2. Do not modify anything else

- Do **not** remove the `git` or `database` configs. Other tests (e.g. `CreateTestStorageManager` at line 311) still use `git`, and the `database` config is exercised by Docker integration tests.
- Do **not** change `storage_helper.go` — the provider switch is intentional per the PR design.
- Do **not** change `ValidateStorageProvider` behavior for flatfile/sqlite. They should pass availability + all four `Create*Store` sub-tests cleanly once configs exist.

### 3. Validation (all three must pass before pushing)

```bash
# Primary: the failing test must pass
go test ./pkg/testing/storage/... -run TestStorageTestFixture_AllProviders -v

# Broader: the whole testing package
go test ./pkg/testing/... -v

# Full local CI parity
make test-complete
```

Verify in the `-v` output that you see sub-tests for `provider_flatfile` and `provider_sqlite` with `--- PASS` (not SKIP, not FAIL). A SKIP here would indicate the provider's `Available()` check failed — investigate rather than accept.

### 4. Architecture check

```bash
make check-architecture
```

Must remain green. The fix does not introduce any provider imports into business logic, so this should stay clean.

---

## Scope Boundaries

- Target branch stays `develop`.
- No new commits or changes outside `pkg/testing/storage/fixtures.go`.
- Commit message: `test(storage): add flatfile and sqlite fixture configs for OSS composite (Issue #692)`
- Push to the existing branch `feature/story-692-agent` — do not open a new PR.

Once pushed, CI should turn green and this PR will be re-reviewed automatically.

--- a/features/workflow/trigger/manager_test.go
+++ b/features/workflow/trigger/manager_test.go
@@ -1074,3 +1074,27 @@ func TestTriggerManagerImpl_ExecuteTrigger(t *testing.T) {
 
 	mockWorkflowTrigger.AssertExpectations(t)
 }
+
+func TestTriggerManagerImpl_NilStoragePersistence(t *testing.T) {
+	// When storage is nil (e.g. composite OSS manager where GetProvider returns nil),
+	// save and delete should be no-ops, not panics.
+	manager := &TriggerManagerImpl{
+		storage:    nil,
+		triggers:   make(map[string]*Trigger),
+		executions: make(map[string]*TriggerExecution),
+		logger:     logging.NewNoopLogger(),
+	}
+
+	ctx := context.Background()
+	trigger := &Trigger{ID: "t-nil-storage", Name: "nil-storage-trigger"}
+
+	t.Run("saveTriggerToStorage returns nil when storage is nil", func(t *testing.T) {
+		err := manager.saveTriggerToStorage(ctx, trigger)
+		assert.NoError(t, err, "saveTriggerToStorage must not error when storage is nil")
+	})
+
+	t.Run("deleteTriggerFromStorage returns nil when storage is nil", func(t *testing.T) {
+		err := manager.deleteTriggerFromStorage(ctx, trigger.ID)
+		assert.NoError(t, err, "deleteTriggerFromStorage must not error when storage is nil")
+	})
+}

--- a/features/workflow/trigger/manager_test.go
+++ b/features/workflow/trigger/manager_test.go
@@ -1082,7 +1082,7 @@ func TestTriggerManagerImpl_NilStoragePersistence(t *testing.T) {
 		storage:    nil,
 		triggers:   make(map[string]*Trigger),
 		executions: make(map[string]*TriggerExecution),
-		logger:     logging.NewNoopLogger(),
+		logger:     logging.ForModule("workflow.trigger.manager.test"),
 	}
 
 	ctx := context.Background()

--- a/pkg/storage/interfaces/provider_test.go
+++ b/pkg/storage/interfaces/provider_test.go
@@ -984,6 +984,158 @@ func (m *MockOSSProvider) CreateRuntimeStore(_ map[string]interface{}) (RuntimeS
 	return nil, ErrNotSupported
 }
 
+// MockOSSProviderWithError is an interface stub that returns an error from a designated Create* method.
+// It is used to test that CreateOSSStorageManager propagates store-creation errors correctly.
+// Real providers cannot be used here because pkg/storage/providers/* imports this package
+// (pkg/storage/interfaces), which would create an import cycle.
+type MockOSSProviderWithError struct {
+	providerName string
+	failMethod   string // name of the Create* method that should fail
+}
+
+func (m *MockOSSProviderWithError) Name() string               { return m.providerName }
+func (m *MockOSSProviderWithError) Description() string        { return "error mock" }
+func (m *MockOSSProviderWithError) GetVersion() string         { return "1.0.0" }
+func (m *MockOSSProviderWithError) Available() (bool, error)   { return true, nil }
+func (m *MockOSSProviderWithError) GetCapabilities() ProviderCapabilities {
+	return ProviderCapabilities{}
+}
+
+func (m *MockOSSProviderWithError) mayFail(method string) error {
+	if m.failMethod == method {
+		return fmt.Errorf("injected %s failure", method)
+	}
+	return nil
+}
+
+func (m *MockOSSProviderWithError) CreateConfigStore(_ map[string]interface{}) (ConfigStore, error) {
+	if err := m.mayFail("CreateConfigStore"); err != nil {
+		return nil, err
+	}
+	return &MockConfigStore{}, nil
+}
+func (m *MockOSSProviderWithError) CreateAuditStore(_ map[string]interface{}) (AuditStore, error) {
+	if err := m.mayFail("CreateAuditStore"); err != nil {
+		return nil, err
+	}
+	return &MockAuditStore{}, nil
+}
+func (m *MockOSSProviderWithError) CreateStewardStore(_ map[string]interface{}) (StewardStore, error) {
+	if err := m.mayFail("CreateStewardStore"); err != nil {
+		return nil, err
+	}
+	return &MockStewardStore{}, nil
+}
+func (m *MockOSSProviderWithError) CreateRBACStore(_ map[string]interface{}) (RBACStore, error) {
+	if err := m.mayFail("CreateRBACStore"); err != nil {
+		return nil, err
+	}
+	return &MockRBACStore{}, nil
+}
+func (m *MockOSSProviderWithError) CreateTenantStore(_ map[string]interface{}) (TenantStore, error) {
+	if err := m.mayFail("CreateTenantStore"); err != nil {
+		return nil, err
+	}
+	return &MockTenantStore{}, nil
+}
+func (m *MockOSSProviderWithError) CreateClientTenantStore(_ map[string]interface{}) (ClientTenantStore, error) {
+	if err := m.mayFail("CreateClientTenantStore"); err != nil {
+		return nil, err
+	}
+	return &MockClientTenantStore{}, nil
+}
+func (m *MockOSSProviderWithError) CreateRegistrationTokenStore(_ map[string]interface{}) (RegistrationTokenStore, error) {
+	if err := m.mayFail("CreateRegistrationTokenStore"); err != nil {
+		return nil, err
+	}
+	return &MockRegistrationTokenStore{}, nil
+}
+func (m *MockOSSProviderWithError) CreateSessionStore(_ map[string]interface{}) (SessionStore, error) {
+	if err := m.mayFail("CreateSessionStore"); err != nil {
+		return nil, err
+	}
+	return &MockSessionStore{}, nil
+}
+func (m *MockOSSProviderWithError) CreateCommandStore(_ map[string]interface{}) (CommandStore, error) {
+	if err := m.mayFail("CreateCommandStore"); err != nil {
+		return nil, err
+	}
+	return &MockCommandStore{}, nil
+}
+func (m *MockOSSProviderWithError) CreateRuntimeStore(_ map[string]interface{}) (RuntimeStore, error) {
+	return nil, ErrNotSupported
+}
+
+func TestCreateOSSStorageManager_StoreCreationErrors(t *testing.T) {
+	// Save and clear registry
+	originalProviders := make(map[string]StorageProvider)
+	globalRegistry.mutex.RLock()
+	for name, provider := range globalRegistry.providers {
+		originalProviders[name] = provider
+	}
+	globalRegistry.mutex.RUnlock()
+
+	globalRegistry.mutex.Lock()
+	globalRegistry.providers = make(map[string]StorageProvider)
+	globalRegistry.mutex.Unlock()
+
+	defer func() {
+		globalRegistry.mutex.Lock()
+		globalRegistry.providers = originalProviders
+		globalRegistry.mutex.Unlock()
+	}()
+
+	// Each subtest injects an error from one of the flatfile Create* methods
+	// to verify CreateOSSStorageManager propagates all store-creation errors.
+	flatfileFailures := []string{
+		"CreateConfigStore",
+		"CreateAuditStore",
+		"CreateStewardStore",
+	}
+	sqliteFailures := []string{
+		"CreateRBACStore",
+		"CreateTenantStore",
+		"CreateClientTenantStore",
+		"CreateRegistrationTokenStore",
+		"CreateSessionStore",
+		"CreateCommandStore",
+	}
+
+	for _, failMethod := range flatfileFailures {
+		failMethod := failMethod
+		t.Run("flatfile_"+failMethod+"_returns_error", func(t *testing.T) {
+			globalRegistry.mutex.Lock()
+			globalRegistry.providers = map[string]StorageProvider{
+				"flatfile": &MockOSSProviderWithError{providerName: "flatfile", failMethod: failMethod},
+				"sqlite":   &MockOSSProvider{providerName: "sqlite"},
+			}
+			globalRegistry.mutex.Unlock()
+
+			_, err := CreateOSSStorageManager(t.TempDir(), t.TempDir()+"/test.db")
+			if err == nil {
+				t.Errorf("expected error when flatfile %s fails, got nil", failMethod)
+			}
+		})
+	}
+
+	for _, failMethod := range sqliteFailures {
+		failMethod := failMethod
+		t.Run("sqlite_"+failMethod+"_returns_error", func(t *testing.T) {
+			globalRegistry.mutex.Lock()
+			globalRegistry.providers = map[string]StorageProvider{
+				"flatfile": &MockOSSProvider{providerName: "flatfile"},
+				"sqlite":   &MockOSSProviderWithError{providerName: "sqlite", failMethod: failMethod},
+			}
+			globalRegistry.mutex.Unlock()
+
+			_, err := CreateOSSStorageManager(t.TempDir(), t.TempDir()+"/test.db")
+			if err == nil {
+				t.Errorf("expected error when sqlite %s fails, got nil", failMethod)
+			}
+		})
+	}
+}
+
 func TestUnregisterStorageProvider(t *testing.T) {
 	// Clear registry for test
 	originalProviders := make(map[string]StorageProvider)

--- a/pkg/testing/storage/fixtures.go
+++ b/pkg/testing/storage/fixtures.go
@@ -102,6 +102,8 @@ func NewStorageTestFixture(t *testing.T) *StorageTestFixture {
 	// Create configurations for all storage providers
 	fixture.setupGitConfig(t)
 	fixture.setupDatabaseConfig(t)
+	fixture.setupFlatfileConfig(t)
+	fixture.setupSQLiteConfig(t)
 
 	return fixture
 }
@@ -156,6 +158,37 @@ func (f *StorageTestFixture) setupDatabaseConfig(t *testing.T) {
 			"password": testPassword,
 			"sslmode":  "disable", // For testing only
 		},
+	}
+}
+
+// setupFlatfileConfig creates a flatfile provider configuration backed by a temp dir.
+func (f *StorageTestFixture) setupFlatfileConfig(t *testing.T) {
+	flatfileDir := filepath.Join(f.TempDir, "flatfile-storage")
+	err := os.MkdirAll(flatfileDir, 0755)
+	require.NoError(t, err, "Failed to create flatfile storage directory")
+
+	f.Configs["flatfile"] = &StorageTestConfig{
+		Provider: "flatfile",
+		Config: map[string]interface{}{
+			"root": flatfileDir,
+		},
+		TempDir: flatfileDir,
+	}
+}
+
+// setupSQLiteConfig creates a sqlite provider configuration backed by a temp file.
+// Uses a real file path (not :memory:) to avoid parallel-test issues.
+func (f *StorageTestFixture) setupSQLiteConfig(t *testing.T) {
+	sqliteDir := filepath.Join(f.TempDir, "sqlite-storage")
+	err := os.MkdirAll(sqliteDir, 0755)
+	require.NoError(t, err, "Failed to create sqlite storage directory")
+
+	f.Configs["sqlite"] = &StorageTestConfig{
+		Provider: "sqlite",
+		Config: map[string]interface{}{
+			"path": filepath.Join(sqliteDir, "cfgms-test.db"),
+		},
+		TempDir: sqliteDir,
 	}
 }
 

--- a/pkg/testing/storage/fixtures.go
+++ b/pkg/testing/storage/fixtures.go
@@ -5,10 +5,12 @@
 package storage
 
 import (
+	"errors"
 	"fmt"
 	"os"
 	"path/filepath"
 	"strconv"
+	"strings"
 	"testing"
 	"time"
 
@@ -18,6 +20,25 @@ import (
 	"github.com/cfgis/cfgms/pkg/storage/interfaces"
 	"github.com/cfgis/cfgms/pkg/testutil"
 )
+
+// isUnsupportedStoreError reports whether err signals that a provider does not
+// implement the requested store type. Post-ADR-003, providers are partitioned
+// by data tier (flatfile: config/audit only; sqlite: business data only), so
+// calling every Create*Store on every provider is expected to yield these
+// errors for out-of-tier combinations. The interfaces package exports
+// ErrNotSupported; individual providers (flatfile) define their own sentinel
+// with the same meaning. Both contain "operation not supported" in their
+// message, so a substring fallback catches provider-local sentinels that do
+// not wrap the canonical one.
+func isUnsupportedStoreError(err error) bool {
+	if err == nil {
+		return false
+	}
+	if errors.Is(err, interfaces.ErrNotSupported) {
+		return true
+	}
+	return strings.Contains(err.Error(), "operation not supported")
+}
 
 // isInfrastructureRequired determines if infrastructure should be available
 // Returns true in CI environments or when Docker/infrastructure is explicitly enabled
@@ -260,6 +281,10 @@ func (f *StorageTestFixture) ValidateStorageProvider(t *testing.T, providerName 
 
 	t.Run(fmt.Sprintf("provider_%s_client_tenant_store", providerName), func(t *testing.T) {
 		store, err := provider.CreateClientTenantStore(testConfig.Config)
+		if isUnsupportedStoreError(err) {
+			t.Skipf("provider %q does not implement ClientTenantStore (ADR-003 tier boundary)", providerName)
+			return
+		}
 		if err != nil && providerName == "database" {
 			requireInfrastructureOrSkip(t, err, "Database provider")
 			return
@@ -274,6 +299,10 @@ func (f *StorageTestFixture) ValidateStorageProvider(t *testing.T, providerName 
 
 	t.Run(fmt.Sprintf("provider_%s_config_store", providerName), func(t *testing.T) {
 		store, err := provider.CreateConfigStore(testConfig.Config)
+		if isUnsupportedStoreError(err) {
+			t.Skipf("provider %q does not implement ConfigStore (ADR-003 tier boundary)", providerName)
+			return
+		}
 		if err != nil && providerName == "database" {
 			requireInfrastructureOrSkip(t, err, "Database provider")
 			return
@@ -288,6 +317,10 @@ func (f *StorageTestFixture) ValidateStorageProvider(t *testing.T, providerName 
 
 	t.Run(fmt.Sprintf("provider_%s_audit_store", providerName), func(t *testing.T) {
 		store, err := provider.CreateAuditStore(testConfig.Config)
+		if isUnsupportedStoreError(err) {
+			t.Skipf("provider %q does not implement AuditStore (ADR-003 tier boundary)", providerName)
+			return
+		}
 		if err != nil && providerName == "database" {
 			requireInfrastructureOrSkip(t, err, "Database provider")
 			return
@@ -302,6 +335,10 @@ func (f *StorageTestFixture) ValidateStorageProvider(t *testing.T, providerName 
 
 	t.Run(fmt.Sprintf("provider_%s_runtime_store", providerName), func(t *testing.T) {
 		store, err := provider.CreateRuntimeStore(testConfig.Config)
+		if isUnsupportedStoreError(err) {
+			t.Skipf("provider %q does not implement RuntimeStore (ADR-003 tier boundary)", providerName)
+			return
+		}
 		if err != nil && providerName == "database" {
 			requireInfrastructureOrSkip(t, err, "Database provider")
 			return

--- a/pkg/testing/storage/fixtures_test.go
+++ b/pkg/testing/storage/fixtures_test.go
@@ -11,7 +11,9 @@ import (
 
 	// Import storage providers for testing
 	_ "github.com/cfgis/cfgms/pkg/storage/providers/database"
+	_ "github.com/cfgis/cfgms/pkg/storage/providers/flatfile"
 	_ "github.com/cfgis/cfgms/pkg/storage/providers/git"
+	_ "github.com/cfgis/cfgms/pkg/storage/providers/sqlite"
 )
 
 func TestStorageTestFixture_Creation(t *testing.T) {
@@ -46,6 +48,20 @@ func TestStorageTestFixture_Creation(t *testing.T) {
 	assert.Contains(t, dbConfig.Config, "database")
 	assert.Contains(t, dbConfig.Config, "username")
 	assert.Contains(t, dbConfig.Config, "password")
+
+	// Verify flatfile configuration
+	flatfileConfig, exists := fixture.GetProviderConfig("flatfile")
+	require.True(t, exists, "Flatfile configuration should exist")
+	require.Equal(t, "flatfile", flatfileConfig.Provider)
+	require.NotNil(t, flatfileConfig.Config, "Flatfile config should not be nil")
+	assert.Contains(t, flatfileConfig.Config, "root")
+
+	// Verify sqlite configuration
+	sqliteConfig, exists := fixture.GetProviderConfig("sqlite")
+	require.True(t, exists, "SQLite configuration should exist")
+	require.Equal(t, "sqlite", sqliteConfig.Provider)
+	require.NotNil(t, sqliteConfig.Config, "SQLite config should not be nil")
+	assert.Contains(t, sqliteConfig.Config, "path")
 }
 
 func TestStorageTestFixture_ControllerConfig(t *testing.T) {

